### DISCUSSION TITLE
Heterogenous Graph Features and Choice of Model

### DISCUSSION THREAD

**Discussion by karan96** [pyg-team/pytorch_geometric#5825] [Q&A] [answered]
Hello all,

I am pretty new to pyg and still getting accustomed to all the methods and models pyg has to offer. My task in hand is to create a heterogeneous graph with two nodes `{'location', 'experts'}`  where locations are mapped from 1 to N for a list of countries C, and experts are mapped from 1 to M from a list of experts E where each expert needs to have an attribute of `{'skills'}`.  A typical attribute for one such node from the experts set looks like this set where each expert is separated by a slash: - `{H01L41/297B41J2/14233H01L41/0471B41J2}`.  I need to create a hetero graph for link prediction but I am not sure how and which model to begin with. With a little bit of study I think GRAPHSAGE should be the choice of model for me. But the tutorial on how to load the data from csv files to Heterodata class are not very good in pyg. I am not sure how each attribute of skill should like, whether it can be a string like above? Or does it need to be a row of sparse matrix converted into tensor and then use each tensor as an attribute to one of the nodes of experts set.  Currently the dataset looks like this:  -

`  data = Data()`
  `data.num_nodes = len(loc_mapping) + len(ex_mapping)`
  `data.edge_index = torch.cat((edge_index, rev_edge_index), dim=1)`
`HeteroData(`
`  num_nodes=13697,`
 ` edge_index=[2, 1258926]`
`)`

`Number of experts I have: 13631`
`Number of loc I have: 66`

I am not sure how to add features to expert nodes and in what type? Plus what model should I be using? I was following this tutorial 
https://medium.com/stanford-cs224w/disease-gene-interactions-with-graph-neural-networks-and-graph-autoencoders-bfae2c28f5ec

and they use GAE wrapping a GCNEncoder model which gives me a constant 0.5 accuracy throughout all the epochs, and all the tensor values coming out of the model were 0 which means this is not working. I changed the GCNEncoder to GNN, using recon loss the AUC was highly fluctuating ranging from 0.10 to 0.9. I ran 20 epochs 100 times to notice the average AUC and it resulted in 0.53 which is highly low considering this model was trained just on node data without any attribute.
Here are my questions: - 
1. How should I add Attributes to experts, and in which format? 
2. Which model and loss function to use in this scenario? 
3. How to increase model accuracy?

Any help would be greatly appreciated. I have the dataset with me in case anyone wants to reproduce.

Best,
Karan

**Comment by rusty1s** [comment, marked as answer]
I just pushed this [colab notebook](https://colab.research.google.com/drive/1xpzn1Nvai1ygd_P5Yambc_oe4VBPK_ZT?usp=sharing) to [our documentation](https://pytorch-geometric.readthedocs.io/en/latest/notes/colabs.html) which seems 100% related to all your questions. Please take a look and let me know if it helps you!

**Reply by jai8004** [reply]
Is there any example of a heterogeneous graph with multiple edges that is uses RCGN to capture the link information and then VGAE to predict the new link?

**Reply by rusty1s** [reply]
This is exactly what the colab notebook linked above does :)

**Reply by karan96** [reply]
Hi @rusty1s,
Thank you for your help.  The notebook does looks like something I might need. Before trying it on my own dataset, 
I tried implementing it for the movielens dataset and faced a couple of problems in getting the dimensions correct. I am using the below settings for transform object: - 
`transform = T.RandomLinkSplit(`
`    num_val=0.10,`
`   num_test=0.10,`
`    disjoint_train_ratio=0.3,`
`    neg_sampling_ratio=0,`
`    add_negative_train_samples=True,`
`    edge_types=("user", "rates", "movie"),`
`    rev_edge_types=("movie", "rev_rates", "user"),`
`)`

A. -> Let me know if these parameters are correct.
B. -> The errors I am getting is at these following lines: -
`assert val_data["user", "rates", "movie"].edge_label_index.size(1) == 30249`. The value I am getting here is: - `10083`
`val_data["user", "rates", "movie"].edge_label.long().bincount().tolist() == [20166, 10083]`. The value I am getting here is: - `[0,10083]`

I couldn't figure out what I am doing wrong.

**Reply by rusty1s** [reply]
It needs to be
```
transform = T.RandomLinkSplit(
num_val=0.10,
num_test=0.10,
disjoint_train_ratio=0.3,
neg_sampling_ratio=2.0,
add_negative_train_samples=False,
edge_types=("user", "rates", "movie"),
rev_edge_types=("movie", "rev_rates", "user"),
)
```

**Reply by karan96** [reply]
Thank you for your continued help @rusty1s . I now understand what the above piece of code is doing. Now when I come onto the LinkNeighborLoader, I am not getting the same output for sampled mini batch. The output I am getting is: - 
`Sampled mini-batch:`
`===================`
`HeteroData(`
`  user={ node_id=[609] },`
`  movie={`
`    node_id=[2699],`
`    x=[2699, 20]`
`  },`
`  (user, rates, movie)={`
`    edge_index=[2, 17058],`
`    edge_label=[384],`
`    edge_label_index=[2, 384]`
`  },`
`  (movie, rev_rates, user)={ edge_index=[2, 7742] }`
`)`

The parameters I am using are: - 

`train_loader = LinkNeighborLoader(`
`    data=train_data,`
`    num_neighbors=[20, 10],`
`    neg_sampling_ratio=2,`
`    edge_label_index=(("user", "rates", "movie"), edge_label_index),`
`    edge_label=edge_label,`
`    batch_size=128,`
`    shuffle=True,`
`)`

Which parameter setting is wrong in my case? Also, how did we decide on what number of neighbors we should sample in each iteration?

**Reply by rusty1s** [reply]
This looks correct and should pass the test. Doesn't it?

**Reply by karan96** [reply]
@rusty1s  No this didn't work, the output I am getting is 
`HeteroData(`
`  user={ node_id=[608] },`
`  movie={`
`    node_id=[2706],`
`    x=[2706, 20]`
`  },`
`  (user, rates, movie)={`
`    edge_index=[2, 16749],`
`    edge_label=[384],`
`    edge_label_index=[2, 384]`
`  },`
`  (movie, rev_rates, user)={ edge_index=[2, 7655] }`
`)`
It is also missing `input_id=[128]`. 
Upon running the model further I am getting loss in negative: - 
`100%|██████████| 190/190 [00:12<00:00, 14.79it/s]`
`  0%|          | 0/190 [00:00<?, ?it/s]Epoch: 001, Loss: -9813.9184`
`100%|██████████| 190/190 [00:12<00:00, 14.88it/s]`
`  0%|          | 0/190 [00:00<?, ?it/s]Epoch: 002, Loss: -10726193.4348`
`100%|██████████| 190/190 [00:11<00:00, 16.01it/s]`
`  0%|          | 0/190 [00:00<?, ?it/s]Epoch: 003, Loss: -396842080.4694`
`100%|██████████| 190/190 [00:11<00:00, 16.73it/s]`
`  0%|          | 0/190 [00:00<?, ?it/s]Epoch: 004, Loss: -4201236530.0131`
`100%|██████████| 190/190 [00:11<00:00, 15.91it/s]`
`Epoch: 005, Loss: -22461009673.3193`

The training code I used is as follows: - 
`sampled_data.to(device)`
`pred = model.forward(sampled_data)`
`loss = F.binary_cross_entropy_with_logits(pred, sampled_data["user", "rates", "movie"].edge_label)`

**Reply by rusty1s** [reply]
Can you install PyG from master (as outlined in the `README.md`)? AFAIK, the notebook expects PyG from master right now.

**Reply by karan96** [reply]
@rusty1s  I did try to install the PyG as per `README.md`  using the command ` conda install pyg -c pyg` on a new environment with a fresh new installation of `torch==1.12.1` but it still didn't solve the above two issues of: - `Assertion Error: assert sampled_data["user", "rates", "movie"].edge_label.max() == 1` and loss being highly negative.

**Reply by rusty1s** [reply]
For installing from master, you need to follow [these](https://github.com/pyg-team/pytorch_geometric#nightly-and-master) instructions. Conda will use the latest release.
